src/engine/interop/specification.md

```diff
@@ -0,0 +1,151 @@
+
+#### Specification
+
+1. Given provided field value set client software **SHOULD** build the initial version of the payload which has an empty transaction set.
+
+2. Client software **SHOULD** start the process of updating the payload. The strategy of this process is implementation dependent. The default strategy would be to keep the transaction set up-to-date with the state of local mempool.
+
+3. Client software **SHOULD** stop the updating process either by finishing to serve the [**`engine_getPayload`**](#engine_getPayload) call with the same `payloadId` value or when [`SECONDS_PER_SLOT`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#time-parameters-1) (currently set to 12 in the Mainnet configuration) seconds have passed since the point in time identified by the `timestamp` parameter.
+
+4. Client software **MUST** set the payload field values according to the set of parameters passed in the call to this method with exception for the `feeRecipient`. The `coinbase` field value **MAY** stem from what is specified by the `feeRecipient` parameter.
+
+5. Client software **SHOULD** respond with `2: Action not allowed` error if the sync process is in progress.
+
+6. Client software **SHOULD** respond with `4: Unknown block` error if the parent block is unknown.
+
+### engine_getPayload
+
+#### Parameters
+1. `payloadId`: `QUANTITY`, 64 Bits - Identifier of the payload building process
+
+#### Returns
+`Object|Error` - Either instance of [`ExecutionPayload`](#ExecutionPayload) or an error
+
+#### Specification
+
+1. Given the `payloadId` client software **MUST** respond with the most recent version of the payload that is available in the corresponding building process at the time of receiving the call.
+
+2. The call **MUST** be responded with `5: Unavailable payload` error if the building process identified by the `payloadId` doesn't exist.
+
+3. Client software **MAY** stop the corresponding building process after serving this call.
+
+### engine_executePayload
+
+#### Parameters
+1. `Object` - Instance of [`ExecutionPayload`](#ExecutionPayload)
+
+#### Returns
+`Object` - Response object:
+1. `status`: `String` - the result of the payload execution:
+  - `VALID` - given payload is valid
+  - `INVALID` - given payload is invalid
+  - `SYNCING` - sync process is in progress
+
+#### Specification
+
+1. Client software **MUST** validate the payload according to the execution environment rule set with modifications to this rule set defined in the [Block Validity](https://eips.ethereum.org/EIPS/eip-3675#block-validity) section of [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#specification) and respond with the validation result.
+
+2. Client software **MUST** defer persisting a valid payload until the corresponding [**`engine_consensusValidated`**](#engine_consensusValidated) message deems the payload valid with respect to the proof-of-stake consensus rules.
```

Comment:
What is the rationale for explicitly announcing to the execution engine that the consensus is validated vs. consensus _invalidated_ and optimistically accepting the block?

Reply:
Is it because you want to avoid the situation where the consensus client is still validating the block and the execution engine has already validated the payload and it returns the block as the latest block via the RPC, then the block is invalidated by the consensus client? Causing the execution engine to return an invalid block to RPC consumers?

Reply:
by leaving a call for both saying INVALID or VALID leaves open the design landscape on execution-layer. They can either optimistically accept (only to do the work to invalidate) or leave it in a limbo to-be-validated cache (only persisting at call to VALID).

I think from a high level spec standpoint (i.e. EIP 3675), it certainly makes sense to be explicit. I think it natural to mirror that here.

Reply:
There should be a clear point in time where the block is considered as accepted and can be served in the wire and used to serve JSON-RPC API requests.

One of the potential design moves is to remove the `status` field from this call and do not send this message for invalid blocks at all, then a block should be considered as invalid by default until this message is received. But in this case execution client would likely have to cache blocks that hasn't received this message for some time, say the clear point of pruning these blocks would be finalizing the checkpoint that is further in the future than blocks that haven't been validated. This may be suboptimal

Reply:
Makes sense.